### DISCUSSION TITLE
Flow Control, ISO 15765-2, or ISO-TP

### DISCUSSION THREAD

**Discussion by scytacki** [meatpiHQ/wican-fw#35] [General]
Looking at the code I don't see any support for CAN flow control.
https://en.wikipedia.org/wiki/ISO_15765-2

I'm just learning all of this stuff, so I might be missing something, but...

From what I can tell, even a basic message like `0902` to request the vin, requires a flow control response before the ECU would send the additional frames/messages.  As far as I can tell the ESP TWAI API doesn't have built in support for this. So my guess is that wican-fw would have to implement this itself. 

At this point I'm only familiar with the ELM327 protocol/chip and how it handles flow control. I'd guess the other protocols have some kind of support for it too. At least with ELM327 by default it tries to handle the flow control for you, and then there are options to override what it does by default.

**Comment by meatpiHQ** [comment]
@scytacki Unless I'm mistaken, flow control is always handled by firmware and not handled by the chip it self. The ELM327 firmware handles the flow control, it can also be implemented in wican-fw, if you're interested in contributing to the project and implementing this feature, a good place to start is the ELM327 [datasheet](https://www.elmelectronics.com/wp-content/uploads/2016/07/ELM327DS.pdf)

**Comment by n135c10r** [comment]
@scytacki im new too in all this CAN related topic, but did you saw https://github.com/brendan-w/python-OBD or https://github.com/pylessard/python-udsoncan? As far as i understand, meatpi is just a device which we can use as socketcan interface with those libs to play with all these obd and uds requests.

**Reply by scytacki** [reply]
@meatpiHQ thanks, yes I will try to add support for flow control to `elm327.c`  I've started working on that file to handle more of the ELM commands including the `AT CP` command. I'll put up a draft PR soon so you can see where I'm headed.  Later, it might be useful to add some of the flow control code to `can.c`, but that depends on how the other protocols deal with it.  Which leads to...

@przemek-n I haven't looked at those libraries closely, neither mentions flow control in their readme. I'm currently focused on ELM327 because that is used by Carscanner as well as A Better Route Planner. The ELM327 protocol has options/commands to override the default flow control response. And Carscanner uses these options/commands to access some of the information in the car.  My guess is that the flow control response has to be sent quickly back to the car, so there might be too much latency if the first frame message had to go over the protocol to the client (via wifi or bluetooth) and then the flow control response had to come back (via wifi or bluetooth).  So my guess is that most protocols work the way ELM327 does: 
- the device (in this case WiCAN) should have a default flow control response
- the protocol provides a way to override this response, which the client does before it sends the actual CAN message to the car. This way the device can respond to the first frame immediately without having to do a round trip to the client.
